holopy/inference/prior.py

```diff
@@ -295,21 +289,66 @@ def sample(self, size=None):
             val[out] = super().sample(len(out[0]))
         return val
 
-    def _add(self, value):
-        return BoundedGaussian(self.mu+value, self.sd, self.lower_bound+value,
-                               self.upper_bound+value, self.name)
 
-    def _multiply(self, value):
-        return BoundedGaussian(self.mu*value, self.sd*value,
-                               self.lower_bound*value, self.upper_bound*value,
-                               self.name)
+class TransformedPrior(Prior):
+    def __init__(self, transformation, base_prior, name=None):
+        """Composite prior composed of one or more base priors transformed by
+        a function. Note there are no `prob` and `lnprob` methods since those
+        just depend on the probabilities of the underlying base priors.
 
-    def __neg__(self):
-        return BoundedGaussian(-self.mu, self.sd, -self.upper_bound,
-                               -self.lower_bound, self.name)
+        Parameters
+        ----------
+        transformation : func
+            Function to apply to base prior to get transformed value
+        base_prior : Prior object or listlike containing Priors
+            Values to be passed into transformation function
+        name : string or None, optional
+            The name of the parameter.
+        """
+        self.base_prior = tuple(ensure_listlike(base_prior))
+        if callable(transformation):
+            self.transformation = transformation
+        else:
+            msg = 'transformation must be function of one or more base priors'
+            raise TypeError(msg)
+        self.name = name
+
+    def prob(self, p):
+        msg = "Cannot calculate probability. Use base priors {}.".format(
+            self.base_names)
+        raise NotImplementedError(msg)
+
+    def lnprob(self, p):
+        msg = "Cannot calculate probability. Use base priors {}.".format(
+            self.base_names)
+        raise NotImplementedError(msg)
+
+    def sample(self, size=None):
+        if size is None:
+            repeat = lambda x: x
+        else:
+            repeat = lambda x: np.repeat(x,size)
+        raw_samples = [bp.sample(size) if isinstance(bp, Prior) else repeat(bp)
+                       for bp in self.base_prior]
+        if size is None:
+            return self.transformation(*raw_samples)
+        else:
+            return np.array([self.transformation(*sample_set)
+                             for sample_set in zip(*raw_samples)])
+        return samples
+
+    @property
+    def guess(self):
+        guess_vals = [bp.guess if isinstance(bp, Prior) else bp
+                      for bp in self.base_prior]
+        return self.transformation(*guess_vals)
+
+    @property
+    def map_keys(self):
+        return enumerate(self.base_prior)
 
 
-class ComplexPrior(Prior):
+class ComplexPrior(TransformedPrior):
```

Comment:
Nice!! :+1:

Reply:
Nope, the most important `TransformedPrior` methods (`guess` and `sample`) are inherited directly.

I do override some placeholders (`prob`, `lnprob`, `map_keys`) in `TransformedPrior` with more intelligent behaviour specific to the `ComplexPrior`. I do think there's a future in which other types of transformations get their own classes (e.g. arctan or other trig), and they would need to override these defaults as well. Let me know if there's a better way to do this.

Finally, there are some properties (`real` and `imag`) that are just convenience ways of accessing `base_prior` components that also maintain backwards compatibility. Actually `map_keys` was added to maintain backwards compatibility as well. I initially had `Model` just follow the `TransformedPrior` behaviour without asking the object, but that didn't match previous parameter names for `ComplexPrior` objects.

Reply:
:+1: